File: l3/z2/lib/GF1234567891.java
package lib;

import java.util.Objects;

public class GF1234567891 implements Field<GF1234567891> {
    protected long value;
    private static final long CHARACTERISTIC = 1234567891;

    public GF1234567891() {
        this.value = 0;
    }

    public GF1234567891(long val) {
        this.value = ((val % CHARACTERISTIC) + CHARACTERISTIC) % CHARACTERISTIC;
    }

    public GF1234567891(GF1234567891 other) {
        this.value = other.value;
    }

    private static long mod(long x)
    {
        return ((x % CHARACTERISTIC) + CHARACTERISTIC) % CHARACTERISTIC;
    }

    @Override
    public long getCharacteristic() {
        return CHARACTERISTIC;
    }

    @Override
    public long getValue() {
        return value;
    }

    private static long multiplicativeInverse(long a) {
        long m0 = CHARACTERISTIC, t, q;
        long x0 = 0, x1 = 1;
        long mod = CHARACTERISTIC;
        if (mod == 1) return 0;
        while (a > 1) {
            q = a / mod;
            t = mod;
            mod = a % mod;
            a = t;
            t = x0;
            x0 = x1 - q * x0;
            x1 = t;
        }
        if (x1 < 0) x1 += m0;
        return x1;
    }

    @Override
    public GF1234567891 add(GF1234567891 rhs) {
        return new GF1234567891(mod(value + rhs.value));
    }

    @Override
    public GF1234567891 subtract(GF1234567891 rhs) {
        return new GF1234567891(mod(value - rhs.value));
    }

    @Override
    public GF1234567891 multiply(GF1234567891 rhs) {
        return new GF1234567891(mod(value * rhs.value));
    }

    @Override
    public GF1234567891 divide(GF1234567891 rhs) {
        if (rhs.value == 0) {
            throw new ArithmeticException("Division by zero is not defined.");
        }
        return multiply(new GF1234567891(multiplicativeInverse(rhs.value)));
    }

    @Override
    public GF1234567891 multiplicativeInverse() {
        return new GF1234567891(multiplicativeInverse(value));
    }

    @Override
    public GF1234567891 createInstance(long value) {
        return new GF1234567891(value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GF1234567891 that = (GF1234567891) o;
        return value == that.value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }
}

File: l3/z3/GF1234567891.cs
using System;

public class GF1234567891 : IFieldElement<long>
{
    private long value;
    private static readonly long CHARACTERISTIC = 1234577;

    public long Value => value;

    public GF1234567891()
    {
        this.value = 1;
    }

    public GF1234567891(long value)
    {
        this.value = Mod(value);
    }

    private static long Mod(long x)
    {
        return ((x % CHARACTERISTIC) + CHARACTERISTIC) % CHARACTERISTIC;
    }

    private static long MultiplicativeInverse(long a)
    {
        long t = 0, newT = 1;
        long r = CHARACTERISTIC, newR = a;
        while (newR != 0)
        {
            long quotient = r / newR;
            long tempT = t - quotient * newT;
            t = newT;
            newT = tempT;
            long tempR = r - quotient * newR;
            r = newR;
            newR = tempR;
        }
        if (r > 1) throw new ArithmeticException("Inverse does not exist.");
        if (t < 0) t += CHARACTERISTIC;
        return t;
    }

    public IFieldElement<long> Add(IFieldElement<long> rhs)
    {
        return new GF1234567891(Mod(this.value + rhs.Value));
    }

    public IFieldElement<long> Subtract(IFieldElement<long> rhs)
    {
        return new GF1234567891(Mod(this.value - rhs.Value));
    }

    public IFieldElement<long> Multiply(IFieldElement<long> rhs)
    {
        return new GF1234567891(Mod(this.value * rhs.Value));
    }

    public IFieldElement<long> Divide(IFieldElement<long> rhs)
    {
        if (rhs.Value == 0)
            throw new ArithmeticException("Division by zero element.");
        return new GF1234567891(Mod(this.value * MultiplicativeInverse(rhs.Value)));
    }

    public override string ToString()
    {
        return value.ToString();
    }

    public long ToLong()
    {
        return value;
    }

    public long GetCharacteristic()
    {
        return CHARACTERISTIC;
    }
}
